### ISSUE TITLE
CentOS: rpm_packages + osqueryd producing rpmdb corruption and not returning logs

### ISSUE THREAD

**Issue by dreamlibrarian** [facebook/osquery#2623]
TL;DR: rpm_packages table + CentOS is misbehaving vigorously, wondering if it's a known issue.

Afternoon, folks - I'm chasing some bizarre stuff and am hoping I'm not the first reporter. That said, this is my first osquery investigation, and I know this is a new version of the RPM tables, so maybe I just picked an exciting time.

I'm using osqueryd to log 'select * from rpm_packages' on CentOS 6 and 7 boxes. 
The daemon (in --verbose mode) logs a variety of berkeleyDB errors in RPM, claims it got results, but never logs them upstream AFAICT.

On the host, the RPM database is corrupted immediately such that Yum won't run, and I have to run rpmdb --rebuild to get it back to sanity.

The same query from osqueryi produces output; in verbose mode it complains about RPM header flags, but produces a table.

I'm running from the 2.0.0 RPM.

Logs from osqueryi:
14 lines like v:
15:44:53.102205 122265 rpm_packages.cpp:50] Could not get RPM header flag.

Logs from osqueryd: 

I1013 15:19:00.707371 118795 scheduler.cpp:63] Executing scheduled query: RPM status: select * from rpm_packages
error: db6 error(-30969) from dbenv->open: BDB0091 DB_VERSION_MISMATCH: Database environment version mismatch
error: cannot open Packages index using db6 -  (-30969)
error: cannot open Packages database in
error: db6 error(-30969) from dbenv->open: BDB0091 DB_VERSION_MISMATCH: Database environment version mismatch
error: cannot open Packages index using db6 -  (-30969)
error: cannot open Packages database in
I1013 15:19:00.712270 118795 query.cpp:68] Storing initial results for new scheduled query: RPM status

**Comment by theopolis** [comment]
Hi @dreamlibrarian, wow, that sounds bad. Definitely haven't seen this before, but we do statically link Berkeley DB and librpm, so there is potential for DB version/header mismatches.

This is happening on CentOS6 and 7 hosts? Here's my output from `osqueryi` on a CentOS 7 machine:
```
» osqueryi --version
osqueryi version 2.0.0
------------------------------------------------------------
» osqueryi -A os_version                                
+--------------+-------+-------+-------+-------+
| name         | major | minor | patch | build |
+--------------+-------+-------+-------+-------+
| CentOS Linux | 7     | 2     | 1511  |       |
+--------------+-------+-------+-------+-------+
------------------------------------------------------------
» osqueryi "select count(1) from rpm_packages" --verbose
I1014 09:06:26.648908 145911 init.cpp:365] osquery initialized [version=2.0.0]
I1014 09:06:26.649746 145914 interface.cpp:307] Extension manager service starting: /home/reed/.osquery/shell.em
I1014 09:06:26.649799 145911 init.cpp:615] Error reading config: config file does not exist: /etc/osquery/osquery.conf
I1014 09:06:27.161348 145911 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:06:27.161393 145911 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:06:27.161412 145911 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:06:27.161442 145911 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:06:27.161459 145911 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:06:27.161476 145911 rpm_packages.cpp:50] Could not get RPM header flag.
+----------+
| count(1) |
+----------+
| 1619     |
+----------+
```

**Comment by theopolis** [comment]
Also, `rpm -qa | wc -l` shows 1619 too.

**Comment by theopolis** [comment]
Here's what I see on a CentOS6 host:
```
~  » osqueryi -A os_version                                
+--------+-------+-------+-------+-------+
| name   | major | minor | patch | build |
+--------+-------+-------+-------+-------+
| CentOS | 6     | 8     | 0     |       |
+--------+-------+-------+-------+-------+
------------------------------------------------------------
~  » osqueryi --version                                    
osqueryi version 2.0.0
------------------------------------------------------------
~  » rpm -qa | wc -l                                       
1359
------------------------------------------------------------
~  » osqueryi --verbose "select count(1) from rpm_packages"
I1014 09:14:57.425277 3975620 init.cpp:365] osquery initialized [version=2.0.0]
I1014 09:14:57.425967 3975624 interface.cpp:307] Extension manager service starting: /home/reed/.osquery/shell.em
error: unrecognized db option: "hash" ignored.
error: unrecognized db option: "tmpdir" ignored.
error: unrecognized db option: "create" ignored.
error: unrecognized db option: "mpool" ignored.
error: unrecognized db option: "perms" ignored.
error: unrecognized db option: "hash" ignored.
error: unrecognized db option: "tmpdir" ignored.
error: unrecognized db option: "create" ignored.
error: unrecognized db option: "mpool" ignored.
error: unrecognized db option: "perms" ignored.
I1014 09:14:57.799026 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799043 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799057 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799059 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799069 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799072 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799080 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799083 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799093 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799094 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799103 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
I1014 09:14:57.799106 3975620 rpm_packages.cpp:50] Could not get RPM header flag.
+----------+
| count(1) |
+----------+
| 1359     |
+----------+
----------------------------------------------------
```

**Comment by theopolis** [comment]
I ran the same query in the `osqueryd` schedule on these hosts and cannot reproduce the DB corruption issue. 

On the CentOS6 host:
```
~  » rpm -qa | grep -i "db4"
db4-cxx-4.7.25-20.el6_7.x86_64
db4-4.7.25-20.el6_7.x86_64
compat-db43-4.3.29-15.el6.x86_64
db4-utils-4.7.25-20.el6_7.x86_64
db4-devel-4.7.25-20.el6_7.x86_64
~  » rpm -qa | grep -i "rpm"
rpm-libs-4.8.0-55.el6.x86_64
rpm-build-4.8.0-55.el6.x86_64
rpm-4.8.0-55.el6.x86_64
rpm-python-4.8.0-55.el6.x86_64
redhat-rpm-config-9.0.3-51.el6.centos.noarch
```

Is there any differences between our setups WRT installed versions of `db` or `rpm`?

**Comment by dreamlibrarian** [comment]
Bloody hell. I suppose it was massively optimistic to hope that this'd be easily reproducible. Actually, glad I'm an edge case.

I missed this this morning, will have to collect package info and see if something bizarre has slipped into my env package-wise.

I was, of course totally sure that db5 (on c7) vs db6 (linked in osqueryi/d) was the culprit here. It's good to know there's something more sinister afoot; I'll get the package info up to validate and see if I can't deepen my knowledge of Berkeley DB enough to make sense of the explosion I'm seeing.

**Comment by theopolis** [comment]
Perfect, thanks for the help debugging!

**Comment by dreamlibrarian** [comment]
Okay, looking closer, 'Could not get RPM header flag' seems to be less panic inducing on its own than I thought (I was lumping it in with the corruption I was seeing).

So... it looks like what I had was some pernicious DB corruption that wasn't resolved by 'rpm --rebuilddb', and happened to be on multiple systems. 

Therefore my starting point (assuming noncorruption) was incorrect. After bludgeoning the test hosts until db_verify stopped emitting hatred, I'm running cleanly (other than the header flag complaint above)

I'm going to close this out and write a new bug (see #2632) for the -actual- issue I saw, which is, given RPM db corruption, queries don't release properly and eventually deadlock the daemon.
